joblib/hashing.py

```diff
@@ -231,8 +242,14 @@ def save(self, obj):
             # To prevent the hash from being sensitive to this, we use
             # .descr which is a full (and never interned) description of
             # the array dtype according to the numpy doc.
-            klass = obj.__class__
-            obj = (klass, ('HASHED', obj.descr))
+            if np_version < "1.20":
+                # Backward compat to avoid breaking old hashes
+                klass = obj.__class__
+                obj = (klass, ('HASHED', obj.descr))
+            else:
```

Comment:
`.descr` is not very suitable for hashing dtypes. For instance, `dtype(int)` and `dtype([('f1', int)])` have the same descr but are otherwise quite different.

In my opinion, there is no problem for which `descr` is an appropriate solution.

Reply:
Thanks for the feedback @eric-wieser. Do you have a better suggestion for hashing a dtype instance? Is the builtin Python `hash(np.float32)` is not stable across Python interpreter restarts.

Reply:
Maybe hashing `(obj.descr, obj.fields)` would be enough?

Reply:
> ~~Is~~ the builtin Python hash(np.dtype('float64')) is not stable across Python interpreter restarts.

I assume you meant this as a statement, not a question.

As far as I remember, even `hash("float")` is not stable across interpreter restarts.

Reply:
Indeed :) My question is more, which publicly accessible attributes of a dtype object can we use to build a deterministic hash (typically use to build cache keys)? Is the `(obj.descr, obj.fields)` pair enough to uniquely identify the dtype object?

Reply:
We currently use `(kind, byteorder, flags, itemsize, alignment)`, but that actually disregards metadata...  It also is pretty bad when it comes to generalization to possible new parametric dtypes, even *if* "kind" is defined as something not completely wrong.  Why can't you hash the pickle? I don't like that unpickling currently does not give you the singleton for `float64`, etc. but I am not actually sure that is a big issue on its own.  I guess its not persistent for multiple NumPy versions, but is that a problem joblib tries to work around?

Reply:
> We currently use `(kind, byteorder, flags, itemsize, alignment)`, but that actually disregards metadata

I cannot work out where `dtype.__hash__` is actually implemented in numpy - it seems we leave `tp_hash` empty...

Reply:
its implemented in `hashdescr.c` typically using the function `_array_descr_builtin`

Reply:
Sorry, about the dtype-meta, I did not touch anything around hashes yet.  That might be an annoying bit to get right (although, I guess the downstream implementer will have to wade through that mainly).  Since the current hashing really only works for builtin types (obviously).  Even existing user dtypes are a bit shady, although they can hope to give a unique `kind` (its one character though currently).

The real problem are parametric dtypes of course, and there it is obviously impossible to provide a reasonable default hash.  Similar things go for pickling the user DType class will have to handle it, at least normally.  Preferably in a way which actually preserves singleton instances (unlike ours right now).  I am a bit worried with all of this about loading a new pickle with an old NumPy versions though.

Reply:
>  I guess its not persistent for multiple NumPy versions, but is that a problem joblib tries to work around?

We do not officially give a guarantee to be able to load old pickles when you upgrade library versions.